### PR TITLE
[Feature] Split data for distributed training

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
For distributed training, we need to split the training/validation/test set among different processes so that each process trains on their own data points.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/distributed/dist_graph.py [hunk 1]
```diff
@@ -116,22 +118,50 @@ def _get_graph_from_shared_mem(graph_name):
...
+
+    We need these metadata to construct graph partition book.
+    '''
+    meta = _move_data_to_shared_mem_array(F.tensor([num_nodes, num_edges,
+                                                    num_partitions, part_id]),
+                                          _get_ndata_path(graph_name, 'meta'))
+    node_map = _move_data_to_shared_mem_array(node_map, _get_ndata_path(graph_name, 'node_map'))
+    edge_map = _move_data_to_shared_mem_array(edge_map, _get_edata_path(graph_name, 'edge_map'))
+    return meta, node_map, edge_map
+
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:138]: I think the _move_metadata_to_shared_mem() will be changed by our new rpc code. All the information shared by client and server will be put into ServerState object.

File: python/dgl/distributed/dist_graph.py [hunk 2]
```diff
@@ -116,22 +118,50 @@ def _get_graph_from_shared_mem(graph_name):
...
 def _get_shared_mem_metadata(graph_name):
     ''' Get the metadata of the graph through shared memory.
 
     The metadata includes the number of nodes and the number of edges. In the future,
     we can add more information, especially for heterograph.
     '''
-    shape = (2,)
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:129]: Same with above comment

File: python/dgl/distributed/dist_graph.py [hunk 3]
```diff
@@ -354,6 +382,14 @@ def __init__(self, server_id, server_namebook, num_client, graph_name, conf_file
...
                 self.set_partition_book(name=_get_edata_name(name), partition_book=edge_map)
 
+        # TODO(zhengda) this is temporary solution. We don't need this in the future.
+        self.meta, self.node_map, self.edge_map = _move_metadata_to_shared_mam(graph_name,
+                                                                               num_nodes,
+                                                                               num_edges,
+                                                                               server_id,
+                                                                               num_partitions,
+                                                                               node_map, edge_map)
+
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:392]: same

File: python/dgl/distributed/dist_graph.py [hunk 4]
```diff
@@ -312,10 +342,8 @@ def __init__(self, server_id, server_namebook, num_client, graph_name, conf_file
         print('Server {}: host name: {}, ip: {}'.format(server_id, host_name, host_ip))
 
         self.client_g, node_feats, edge_feats, self.meta = load_partition(conf_file, server_id)
-        num_nodes, num_edges, node_map, edge_map = self.meta
+        num_nodes, num_edges, node_map, edge_map, num_partitions = self.meta
         self.client_g = _copy_graph_to_shared_mem(self.client_g, graph_name)
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:346]: same

File: python/dgl/distributed/dist_graph.py [hunk 5]
```diff
@@ -383,19 +419,12 @@ def __init__(self, server_namebook, graph_name):
         self._client = KVClient(server_namebook=server_namebook)
         self._client.connect()
 
-        self.g = _get_graph_from_shared_mem(graph_name)
-        self.graph_name = graph_name
-        self.meta = F.asnumpy(_get_shared_mem_metadata(graph_name))
+        self._g = _get_graph_from_shared_mem(graph_name)
```
Comment (aksnzhy) [inline python/dgl/distributed/dist_graph.py:422]: same

**Comment by aksnzhy** [comment]
The high-level conner is that all the shared-memory information (metadata, node data, edge data, graph, and also GraphPartitionBook) will be put into ServerState in new rpc code. It seems that many code in DistGraph will also be changed.

File: python/dgl/distributed/dist_graph.py (hunk 1 above)
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:138]: i know that. btw, why is ServerState in the RPC code. RPC should just be used for communication.

**Comment by zheng-da** [comment]
It's inevitable. What is the alternative solution right now?

File: python/dgl/distributed/dist_graph.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/distributed/dist_graph.py:139]: @zheng-da Agree. We can adjust that later.

File: python/dgl/distributed/dist_graph.py [hunk 6]
```diff
@@ -557,3 +599,80 @@ def _get_all_edata_names(self):
                 # Remove the prefix "edge:"
                 edata_names.append(name[5:])
         return edata_names
+
+def _get_part(elements, local_ids):
```
Comment (jermainewang) [inline python/dgl/distributed/dist_graph.py:603]: docstring

File: python/dgl/distributed/dist_graph.py [hunk 7]
```diff
@@ -557,3 +599,80 @@ def _get_all_edata_names(self):
...
+        The rank of the current process
+
+    Returns
+    -------
+    1D-tensor
+        The vector of node Ids that belong to the rank.
+    '''
+    num_nodes = 0
+    for part in partition_book.metadata():
+        num_nodes += part['num_nodes']
```
Comment (jermainewang) [inline python/dgl/distributed/dist_graph.py:639]: I remember there is a `total_num_nodes` field?

File: python/dgl/distributed/dist_graph.py [hunk 8]
```diff
@@ -557,3 +599,80 @@ def _get_all_edata_names(self):
...
+        The rank of the current process
+
+    Returns
+    -------
+    1D-tensor
+        The vector of edge Ids that belong to the rank.
+    '''
+    num_edges = 0
+    for part in partition_book.metadata():
+        num_edges += part['num_edges']
```
Comment (jermainewang) [inline python/dgl/distributed/dist_graph.py:673]: same comment

File: tests/distributed/test_dist_graph_store.py [hunk 9]
```diff
@@ -76,10 +76,21 @@ def run_client(graph_name, barrier, num_nodes, num_edges):
     assert g.node_attr_schemes()['test1'].dtype == F.int32
     assert g.node_attr_schemes()['features'].shape == (1,)
 
+    selected_nodes = np.random.randint(0, 100, size=g.number_of_nodes()) > 30
+    # Test node split
+    local_nids = F.nonzero_1d(g._g.ndata['local_node'])
+    local_nids = g._g.ndata[dgl.NID][local_nids]
```
Comment (jermainewang) [inline tests/distributed/test_dist_graph_store.py:82]: Can we avoid using `g._g` in unittests? We should treat us as users when writing unittests.

File: python/dgl/distributed/dist_graph.py (hunk 7 above)
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:639]: it doesn't seem to have such a field. the metadata is stored in a list of dict. each dict has the information of the partition.

**Pull request merged**